File: LeetCode/merge_nodes_in_between_zeros.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        ListNode *ans, *currAns;
        ans = nullptr;
        
        int sum = 0;
        bool begin = false;
        ListNode *curr = head;
        while(curr) {
            if(!curr->val && !begin) {
                begin = true;
            } else if(!curr->val && begin) {
                if(!ans) {
                    ans = new ListNode(sum);
                    currAns = ans;
                } else {
                    currAns->next = new ListNode(sum);
                    currAns = currAns->next;
                }

                sum = 0;
            }
            
            sum += curr->val;
            curr = curr->next;
        }
        
        return ans;
    }
};

File: LeetCode/merge_nodes_in_between_zeros.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeNodes(head *ListNode) *ListNode {
    var ans *ListNode = nil
    sum, curr, currAns := 0, head.Next, ans

    for curr != nil {
        sum += curr.Val

        if curr.Val == 0 {
            next := &ListNode{
                Val: sum,
                Next: nil,
            }

            if ans == nil {
                ans = next
                currAns = ans
            } else {
                currAns.Next = next
                currAns = currAns.Next
            }

            sum = 0
        }

        curr = curr.Next
    }

    return ans
}
